### DISCUSSION TITLE
ui-avatars - User avatar on collection

### DISCUSSION THREAD

**Discussion by matijarezek** [filamentphp/filament#9903] [Help]
I have a collection of users that are 'collaborators' on ticket model. It it set up as a belongs to many relationship with pivot table.

When I want to display user avatars on table or infoview I cant seem to display user avatars as stacked images.
I think thats because my users dont have avatar collumn in database and so they fallback to default avatar provider which is ui-avatars.

I tried with ->defaultImageUrl but it doesent work with collections as mine. 

` ImageEntry::make('collaborators.avatar')` this ofcourse doesen't work.

Is there any easy solution for this?

**Comment by wickham-c** [comment]
A bit late I know, but I had a similar problem and found a solution [here](https://www.answeroverflow.com/m/1207607817251590164). 

Essentially, you need to:
1. [Create an attribute](https://laravel.com/docs/11.x/eloquent-mutators#defining-an-accessor) on the Collaborator/User model that returns the fallback ui-avatars URL (you could call `\Filament\Facades\Filament::getUserAvatarUrl($this);` to return this, or just use the final line of code in that method if you don't use a database table column to store avatar URLs).
2. Add that new attribute to the Collaborator/User model's `$appends` property, as described [here](https://laravel.com/docs/11.x/eloquent-serialization#appending-values-to-json).
3. Pass that new attribute name as the argument to your `ImageEntry::make()` call.

**Comment by emogeekface** [comment]
@wickham-c and @matijarezek 

I am having the same issue as you, but I'm trying to implement this solution and it just wont work.

I have a `User` Model (with an avatar column in my db), a `Requests` model, and then a `Team` relationship between `User` and `Request` via a pivot table.

When trying to use stacked images to display the requests' team I can get it to display the avatar image if one has been uploaded, but if no image has been set for the user it displays nothing, instead of the ui-avatars fallback URL.

My `User` modal has this function in it:

```
public function getFilamentAvatarUrl(): ?string
    {
        if($this->avatar) {
            return Storage::url($this->avatar);
        } else {
            return 'https://ui-avatars.com/api/?name=' . urlencode(str_replace('-', '', $this->name));
        }
    }
```
which is being called elsewhere via `user.avatar` and works perfectly fine.

My `RequestResource` infolist has

```
ImageEntry::make('team.avatar')
->height(40)
->circular()
->stacked()
```

Which, as I say, does work when an image is uploaded but displays nothing at all if not.

I've tried doing as said above and appending a `team_avatar` value within the `RequestUser` Model, but that seems to get completely ignored. I'm not sure what I'm doing wrong. 

Do any of you have any code you could share to illustrate what you've done to fix this?

**Comment by jinbatsu** [comment]
I am having the same issue as both of you all @matijarezek, @emogeekface, And @wickham-c . And it works for me.
I consider extend **ImageColumn**, and replace the **getImageUrl** function to use my **src** attribute, like this:
```php
class ComColPasFotoStacked extends ImageColumn
{
    protected function setUp(): void {
        parent::setUp();
        $this->myRender();
    }

    private function myRender(): void {
        $this->size(40) // default 40
            ->circular()
            ->stacked()
            ->ring(0) // 0-8, default 3
            ->overlap(4) // 0-8, default 4
            ->toggleable(isToggledHiddenByDefault: false);
    }

    public function getImageUrl(?string $state = null): ?string {
        $user = User::withTrashed()->find($state);
        $state = $user->avatar;
        return $state;
    }
}
```
I'm using Laravolt Avatar, but here just example when using ui-avatar or any external URL.
Here is the Model:
```php
public function getAvatarAttribute(): ?string {
	if($this->avatar) {
            return Storage::url($this->avatar);
        } else {
            return 'https://ui-avatars.com/api/?name=' . urlencode(str_replace('-', '', $this->name));
        }
}
```

And here is how to use in table, get the **id** so that the function will populate as needed:
```php
ComColPasFotoStacked::make('colaborators.user.id')->label('Colaborator')
```
